### PR TITLE
Add Filters to User Portal

### PR DESCRIPTION
**Pull request by admy7**
This PR adds filtering to the catalogue. 

The filters have been designed in a reusable manner. 

Currently, there are 5 filters: Publishers, Formats, Catalogues, Themes, Keywords. 
They can easily be modified / extended.

Please note that the filtering on "formats" is currently not working properly. 2 issues: 
    - The data coming from the FAIR data points harvester has no "format" property
    - The currently existing endpoint does not allow to filter on "format". The payload is as follows: it contains a "resources" property which is an array of objects, which have a "format" property. We may have to carry out modifications on the backend to make this filtering work properly.

A few screenshots: 

Screenshot 2024-01-09 at 09 49 45

Screenshot 2024-01-09 at 09 50 08

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -2,17 +2,9 @@ version: "3"
 
 services:
   frontend:
-    build: 
-      context: .
+    image: ghcr.io/genomicdatainfrastructure/gdi-userportal-frontend:main
```
Comment (inderps) [inline docker-compose.yml:5]: All these prod deploy related docker changes could have been in a separate PR

**Review by nolliia — APPROVED** [review]

File: docker-compose.yml (hunk 1 above)
Comment (brunopacheco1) [reply docker-compose.yml:5]: This `docker-compose.yml` is not prod, I created it to test locally the Dockerfile, to understand if NGINX was getting properly the environment variables.

Anything related to environment deployment is held here: https://github.com/GenomicDataInfrastructure/gdi-userportal-ckan-deployment

Comment (brunopacheco1) [reply docker-compose.yml:5]: @admy7 can you please put revert these changes?

File: src/environment/environment.ts [hunk 2]
```diff
@@ -1,10 +1,6 @@
 export const environment = {
-    // @ts-ignore
-    production: window['env']['production'],
-    // @ts-ignore
```
Comment (brunopacheco1) [inline src/environment/environment.ts:4]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

File: src/environment/environment.prod.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export const environment = {
```
Comment (brunopacheco1) [inline src/environment/environment.prod.ts:1]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

File: Dockerfile [hunk 4]
```diff
@@ -6,15 +6,10 @@ COPY ./ /usr/local/app/
...
 RUN npm install
 
-RUN npm run build
+RUN npm run build --configuration=production
 
 FROM nginx:latest
 
 COPY --from=build /usr/local/app/dist/gdi-userportal-app /usr/share/nginx/html
 
-COPY nginx/ /etc/nginx/
```
Comment (brunopacheco1) [inline Dockerfile:15]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

File: .dockerignore [hunk 5]
```diff
@@ -1,6 +0,0 @@
-.git
-.gitignore
```
Comment (brunopacheco1) [inline .dockerignore:2]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

**Review by brunopacheco1 — CHANGES_REQUESTED** [review]

File: src/assets/env.template.js [hunk 6]
```diff
@@ -1,8 +0,0 @@
-(function (window) {
```
Comment (brunopacheco1) [inline src/assets/env.template.js:1]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

File: src/assets/env.js [hunk 7]
```diff
@@ -1,8 +0,0 @@
-(function (window) {
```
Comment (brunopacheco1) [inline src/assets/env.js:1]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

File: _http/catalogue.http [hunk 8]
```diff
@@ -1 +0,0 @@
-GET http://localhost:4200/api/3/action/package_search?start=0&rows=5
```
Comment (brunopacheco1) [inline _http/catalogue.http:1]: Can you please revert these changes? I believe there was a rebase/merging issue with this branch.

File: src/styles.scss [hunk 9]
```diff
@@ -97,4 +97,8 @@ a {
 
 a:hover {
   color: var(--scampi-blue);
-}
\ No newline at end of file
+}
+
+mat-card {
+  background-color: #7276b118 !important;
```
[the remark this answers is not in the document]
Comment (admy7) [reply src/styles.scss]: Thanks for pointing it out. Based on your recommendation, I created a variable for storing the color (in the global scss file named "styles.scss", and a class in the search result component scss

File: docker-compose.yml (hunk 1 above)
Comment (admy7) [reply docker-compose.yml]: Done

File: src/index.html [hunk 10]
```diff
@@ -9,7 +9,6 @@
   <link rel="icon" type="image/x-icon" href="/assets/favicon/favicon.svg">
   <link rel="preconnect" href="https://fonts.gstatic.com">
   <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
-  <script src="assets/env.js"></script>
```
[the remark this answers is not in the document]
Comment (admy7) [reply src/index.html]: Done

File: src/environment/environment.ts (hunk 2 above)
Comment (admy7) [reply src/environment/environment.ts]: Done

File: src/environment/environment.prod.ts (hunk 3 above)
Comment (admy7) [reply src/environment/environment.prod.ts]: Done

File: nginx/nginx.conf [hunk 11]
```diff
@@ -1,44 +0,0 @@
-worker_processes 1;
-
```
[the remark this answers is not in the document]
Comment (admy7) [reply nginx/nginx.conf]: Done

File: Dockerfile (hunk 4 above)
Comment (admy7) [reply Dockerfile]: Done

File: .dockerignore (hunk 5 above)
Comment (admy7) [reply .dockerignore]: Done

File: src/assets/env.template.js (hunk 6 above)
Comment (admy7) [reply src/assets/env.template.js]: Done

File: src/assets/env.js (hunk 7 above)
Comment (admy7) [reply src/assets/env.js]: Done

File: _http/catalogue.http (hunk 8 above)
Comment (admy7) [reply _http/catalogue.http]: Done

File: src/app/services/ckan.service.ts [hunk 12]
```diff
@@ -2,70 +2,100 @@ import { Injectable } from '@angular/core';
...
-import { Dataset } from '../interfaces/dataset-details.interface';
+import { Dataset } from '../interfaces/dataset-details';
+import { CatalogueDetail } from '../interfaces/catalogue-details';
+import { PartialDataset } from '../interfaces/dataset';
 
 @Injectable({
-  providedIn: 'root'
+  providedIn: 'root',
 })
 export class CkanService {
```
[the remark this answers is not in the document]
Comment (admy7) [reply src/app/services/ckan.service.ts:12]: Totally agreed for the need of this test. I will create a new story for that though, to unblock this PR

File: src/app/components/search-results/search-results.component.ts [hunk 13]
```diff
@@ -2,103 +2,115 @@ import { Component, OnInit } from '@angular/core';
...
+    private ckanService: CkanService,
+    private router: Router
+  ) {}
 
   ngOnInit(): void {
-    this.route.queryParams.subscribe(params => {
-      const query = params['query'];
-      this.loadSearchResults(query);
+    // Retrieve all values for filters (must go through all the pages)
+    this.ckanService.searchDatasets('', '', 0, 1_000_000_000_000).subscribe((data) => {
```
Comment (brunopacheco1) [inline src/app/components/search-results/search-results.component.ts:34]: I wonder if this is correct. There should be better ways to retrieve filter values without the need of querying datasets.

File: src/app/services/ckan.service.ts [hunk 14]
```diff
@@ -2,70 +2,96 @@ import { Injectable } from '@angular/core';
...
           id: item.id,
           title: item.title,
           description: item.notes,
           modified: item.metadata_modified,
-          organizationName: item.organization.title,
-          publisher_name: item.publisher_name
+          organization: item.organization.title,
+          publisher_name: item.publisher_name,
+          tags: item.tags.map((tag: { name: string }) => tag.name),
+          theme: item.theme[0]
```
Comment (brunopacheco1) [inline src/app/services/ckan.service.ts:43]: I though this problem was fixed and this value was retrieve as an array, but if not - this is a string of array of strings - it whould be simpler running  `JSON.parse(intem.theme[0])`, no?

File: src/app/services/ckan.service.ts [hunk 15]
```diff
@@ -2,70 +2,96 @@ import { Injectable } from '@angular/core';
...
-      toArray(),
+      mergeMap((url) => this.getCatalogueDetail(url)),
+      toArray()
     );
   }
 
-  getCatalogueDetail(url: string): Observable<any> {
-    let itemCategory: string = this.ckanToDcat.get(url.split("/").pop()?.split("_")?.[0] ?? "") ?? "";
+  private getCatalogueDetail(url: string): Observable<CatalogueDetail> {
+    let itemCategory: string = CkanService.ckanToDcat.get(url.split('/').pop()!.split('_')[0])!;
```
Comment (brunopacheco1) [inline src/app/services/ckan.service.ts:82]: You parse to URL, to parse it back to key here. Can you simply passa both key and url as args for this function? Or parse to URL inside the function?

File: src/app/interfaces/catalogue-details.ts [hunk 16]
```diff
@@ -0,0 +1,4 @@
+export interface CatalogueDetail {
```
Comment (brunopacheco1) [inline src/app/interfaces/catalogue-details.ts:1]: I wonder if this name is correct. What does it retrieve example? Field values and counter per value?

File: src/app/services/ckan.service.ts [hunk 17]
```diff
@@ -2,70 +2,96 @@ import { Injectable } from '@angular/core';
...
+import { CatalogueDetail } from '../interfaces/catalogue-details';
+import { PartialDataset } from '../interfaces/dataset';
 
 @Injectable({
-  providedIn: 'root'
+  providedIn: 'root',
 })
 export class CkanService {
-  ckanToDcat: Map<string, string> = new Map(Object.entries({ package: "Dataset", organization: "Catalogue", tag: "Keyword", group: "Theme" }));
+  private static readonly ckanToDcat: Map<string, string> = new Map(
```
Comment (brunopacheco1) [inline src/app/services/ckan.service.ts:13]: Can you please rename this variable to upper snake case, as it is static and final?

File: src/app/services/ckan.service.ts [hunk 18]
```diff
@@ -2,70 +2,96 @@ import { Injectable } from '@angular/core';
...
+    Object.entries({
```
Comment (brunopacheco1) [inline src/app/services/ckan.service.ts:14]: Why can you not simply use the object directly? This key value pairs will never change, and the ordering of the keys are not relevant.

**Review by brunopacheco1 — APPROVED** [review]
